crates/evm-ee/src/types.rs

```diff
@@ -0,0 +1,348 @@
+use strata_codec::{Codec, CodecError};
+use strata_ee_acct_types::{EnvResult, ExecBlock, ExecBlockBody, ExecHeader, ExecPartialState};
+
+pub(crate) type Hash = [u8; 32];
+
+/// Partial state for EVM execution using RSP's sparse Merkle Patricia Trie.
+///
+/// This represents a sparse state containing only the accounts and storage slots
+/// that were accessed during execution. It's optimized for zero-knowledge proof
+/// generation where including the full Ethereum state would be prohibitively expensive.
+///
+/// Also includes witness data needed for block execution:
+/// - Contract bytecodes for executing contract calls
+/// - Ancestor block headers for the BLOCKHASH opcode
+#[derive(Clone, Debug)]
+pub struct EvmPartialState {
+    /// The sparse Merkle Patricia Trie state from RSP
+    ethereum_state: EthereumState,
+    /// Contract bytecodes needed for execution
+    /// FIXME: ensure bytecodes and ancestor_header are right to be placed in this struct
+    bytecodes: Vec<Bytecode>,
+    /// Ancestor block headers for BLOCKHASH opcode support
+    ancestor_headers: Vec<Header>,
+}
+
+impl EvmPartialState {
+    /// Creates a new EvmPartialState from an EthereumState with witness data.
+    pub fn new(
+        ethereum_state: EthereumState,
+        bytecodes: Vec<Bytecode>,
+        ancestor_headers: Vec<Header>,
+    ) -> Self {
+        Self {
+            ethereum_state,
+            bytecodes,
+            ancestor_headers,
+        }
+    }
+
+    /// Gets a reference to the underlying EthereumState.
+    pub fn ethereum_state(&self) -> &EthereumState {
+        &self.ethereum_state
+    }
+
+    /// Gets a mutable reference to the underlying EthereumState.
+    pub fn ethereum_state_mut(&mut self) -> &mut EthereumState {
+        &mut self.ethereum_state
+    }
+
+    /// Gets a reference to the bytecodes.
+    pub fn bytecodes(&self) -> &[Bytecode] {
+        &self.bytecodes
+    }
+
+    /// Gets a reference to the ancestor headers.
+    pub fn ancestor_headers(&self) -> &[Header] {
+        &self.ancestor_headers
+    }
+}
+
+impl ExecPartialState for EvmPartialState {
+    fn compute_state_root(&self) -> EnvResult<Hash> {
+        let state_root = self.ethereum_state.state_root();
+        Ok(state_root.into())
+    }
+}
+
+impl Codec for EvmPartialState {
+    fn encode(&self, _enc: &mut impl strata_codec::Encoder) -> Result<(), CodecError> {
+        // TODO: Implement proper encoding for EthereumState
+        // For now, we'll return an error as this needs RSP's serialization
+        Err(CodecError::InvalidVariant(
+            "EthereumState encoding not implemented",
+        ))
+    }
+
+    fn decode(_dec: &mut impl strata_codec::Decoder) -> Result<Self, CodecError> {
+        // TODO: Implement proper decoding for EthereumState
+        // For now, we'll return an error as this needs RSP's deserialization
+        Err(CodecError::InvalidVariant(
+            "EthereumState decoding not implemented",
+        ))
+    }
```

Comment:
Making a note not to forget this impl.

Also, `InvalidVariant` isn't when there's an enum variant discriminator that's not recognized.  Not sure what a more appropriate error variant would be here.